pkg/controller/baremetalhost/baremetalhost_controller.go

```diff
@@ -607,19 +665,43 @@ func (r *ReconcileBareMetalHost) manageHostPower(prov provisioner.Provisioner, i
 		return actionContinue{provResult.RequeueAfter}
 	}
 
+	pendingRebootSince := info.host.Status.Provisioning.PendingRebootSince
+	poweredOnAt := info.host.Status.Provisioning.PoweredOnAt
+
+	desiredPowerOnState := info.host.Spec.Online
+	isInRebootProcess := poweredOnAt.Before(pendingRebootSince) || (!pendingRebootSince.IsZero() && poweredOnAt.IsZero())
+
+	if isInRebootProcess {
+		if !info.host.Status.PoweredOn {
+			if _, hasAnn := info.host.Annotations[rebootAnnotationPrefix]; hasAnn {
+				delete(info.host.Annotations, rebootAnnotationPrefix)
+
+				if err = r.client.Update(context.TODO(), info.host); err != nil {
+					return actionError{errors.Wrap(err, "failed to remove reboot annotation from host")}
+				}
+				return actionContinue{}
+			}
+		}
+
+		if info.host.Status.PoweredOn || hasRebootAnnotation(info.host) {
```

Comment:
I eventually convinced myself that this condition is correct, but it was hard work :D

* If a reboot annotation is still present, we want to force the power off.
* If not, it could be because someone added a suffixless annotation, we saw it and set `pendingRebootSince`, but then someone else removed the annotation. We don't want reboots to be cancellable, so we need to go through with it: as long as the power is still on we want to force it off.

No suggestion, just leaving this here so I don't have to work it out again next time.

Reply:
Maybe an inline comment with the reasoning would be good.

Reply:
That sounds like a very good idea